docs/RFCS/add_sstable.md

```diff
@@ -0,0 +1,119 @@
+- Feature Name: AddSSTable Command
+- Status: draft
+- Start Date: 2017-05-12
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A new kv command to use RocksDB's [IngestExternalFile] (previously called
+`AddFile`) which links a prebuilt SSTable directly into the underlying
+log-structured merge-tree.
+
+# Motivation
+
+`RESTORE` of a backup and bulk `LOAD` of external data both need to put data in
+RocksDB as fast as possible.
+
+Previously, this worked via the `WriteBatch` command, which commits the
+serialized form of a RocksDB batch. This unfortunately involves a high write
+amplification (because of compactions) and keeps a duplicate of the data in the
+raft log. Both of these have contributed to stability issues.
+
+Additionally, `disk max throughput / write amplification` is currently the
+limiting constraint on bulk import speed.
+
+TODO(dan): It's possible that snapshots could reuse all or parts of this work.
+Update this RFC with details.
+
+# Detailed design
+
+A new kv command `AddSSTable` takes an `SSTable` generated by `SstFileWriter` as
+a `[]byte` payload. (Please bikeshed the name. Some alternatives: AddTable,
+IngestExternal, AddFile) The proposer validates both that the encoded keys are
+within the declared span and that the key-value checksums are correct. MVCC
+stats are calculated. The proposer then content addresses the payload and puts
+only this checksum into raft.
+
+Each replica, when applying this raft entry, retrieves the payload (details
+below), writes it to disk in the store directory, verifies the checksum, and
+inserts the file into RockdDB via `IngestExternalFile`.
+
+## [WIP] Payload Trasnfer
+
+Each replica needs the payload to apply the raft command, but by design the
+payload is not passed through the normal raft mechanism. The most delicate part
+of AddSSTable is making sure the payload is present (or retrievable) for this
+step. If it's ever not, the replica will not be able to make progress. If this
+happens to a majority of replicas, the range will not be able to make progress.
```

Comment:
It could make progress if there's another "ingest" command that replaces the current content (i.e. makes the result of an earlier ingest obsolete). Probably a bit tricky to put this into practice though.